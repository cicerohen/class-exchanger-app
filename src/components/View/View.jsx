import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Style from './style/View.scss';

class View extends Component {
  render() {
    const {
      title, subtitle, header, sidebar, footer,
      sidebarIsOpen, closeSidebarFn, children,
    } = this.props;
    return (
      <div className={Style.root}>
        <div className={Style.container}>
          {
            sidebarIsOpen &&
            <div className={Style.sidebarRoot}>
              <div className={Style.sidebarContainer}>
                {sidebar}
              </div>
              <div className={Style.sidebarBackdrop} onClick={closeSidebarFn} />
            </div>
          }
          <div>
            <div className={Style.headerContainer}>
              {header}
            </div>
            <main className={Style.mainContainer}>
              <h2 className={Style.viewTitle}>{title}</h2>
              {
                subtitle && <div className={Style.subtitleContainer}>{subtitle}</div>
              }
              {children}
            </main>
            <div className={Style.footerContainer}>
              {footer}
            </div>
          </div>
        </div>
      </div>
    );
  }
}

View.propTypes = {
  // required
  title: PropTypes.string.isRequired,
  header: PropTypes.node.isRequired,
  sidebar: PropTypes.node.isRequired,
  footer: PropTypes.node.isRequired,
  sidebarIsOpen: PropTypes.bool.isRequired,
  closeSidebarFn: PropTypes.func.isRequired,
  // optional
  subtitle: PropTypes.string,
  children: PropTypes.oneOfType([
    PropTypes.node,
    PropTypes.arrayOf(PropTypes.node),
  ]),
};

View.defaultProps = {
  subtitle: '',
  children: undefined,
};

export default View;
