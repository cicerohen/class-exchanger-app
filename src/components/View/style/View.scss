@import "~scss/utils/_unit";


.mainContainer {
  padding-top: rem-calc(10);
  padding-bottom: rem-calc(10); 
  max-width: 1120px;
  margin: 0 auto;
}

.sidebarRoot {
  position: fixed;
  width: 100%;
  height: 100%;
  z-index: 100;
}

.sidebarContainer {
  max-width: 350px;
  height: 100%;
  position: relative;
  z-index: 100;
  box-shadow: 0 0 5px #000;
}

.sidebarBackdrop {
  width: 100%;
  height: 100%;
  position: absolute;
  background-color: rgba(0,0,0, 0.8);
  top: 0;
  left: 0;
}