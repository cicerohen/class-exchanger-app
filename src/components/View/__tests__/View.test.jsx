import React from 'react';
import { shallow } from 'enzyme';
import View from '../View';

const minimalProps = {
  title: 'View Title',
  header: <div className="header" />,
  sidebar: <div className="sidebar" />,
  footer: <div className="footer" />,
  sidebarIsOpen: false,
  closeSidebarFn: () => {},
};

describe('<View /> Tests', () => {
  it('should render without exploding', () => {
    const wrapper = shallow(<View {...minimalProps} />);
    expect(wrapper).toHaveLength(1);
  });

  it('should render correctly with minimal props', (done) => {
    const wrapper = shallow(<View {...minimalProps}><div>content</div></View>);
    expect(wrapper).toMatchSnapshot();
    done();
  });

  it('should render correctly with subtitle', (done) => {
    const wrapper = shallow(<View {...minimalProps} subtitle="View Subtitle" />);
    expect(wrapper).toMatchSnapshot();
    done();
  });

  it('should render correctly with sidebar opened', (done) => {
    const wrapper = shallow(<View {...minimalProps} sidebarIsOpen />);
    expect(wrapper).toMatchSnapshot();
    done();
  });

  it('should call closeSidebarFn when click on sidebar backdrop', (done) => {
    const closeSidebarFnMock = jest.fn();
    const wrapper = shallow(
      <View
        {...minimalProps}
        sidebarIsOpen
        closeSidebarFn={closeSidebarFnMock}
      />
    );
    const sidebarBackdrop = wrapper.find('.sidebarBackdrop');
    sidebarBackdrop.simulate('click');
    expect(closeSidebarFnMock.mock.calls.length).toBe(1);
    done();
  });
});

