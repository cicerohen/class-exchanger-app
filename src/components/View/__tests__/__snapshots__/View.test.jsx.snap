// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<View /> Tests should render correctly with minimal props 1`] = `
<div
  className="root"
>
  <div
    className="container"
  >
    <div>
      <div
        className="headerContainer"
      >
        <div
          className="header"
        />
      </div>
      <main
        className="mainContainer"
      >
        <h2
          className="viewTitle"
        >
          View Title
        </h2>
        <div>
          content
        </div>
      </main>
      <div
        className="footerContainer"
      >
        <div
          className="footer"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<View /> Tests should render correctly with sidebar opened 1`] = `
<div
  className="root"
>
  <div
    className="container"
  >
    <div
      className="sidebarRoot"
    >
      <div
        className="sidebarContainer"
      >
        <div
          className="sidebar"
        />
      </div>
      <div
        className="sidebarBackdrop"
        onClick={[Function]}
      />
    </div>
    <div>
      <div
        className="headerContainer"
      >
        <div
          className="header"
        />
      </div>
      <main
        className="mainContainer"
      >
        <h2
          className="viewTitle"
        >
          View Title
        </h2>
      </main>
      <div
        className="footerContainer"
      >
        <div
          className="footer"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<View /> Tests should render correctly with subtitle 1`] = `
<div
  className="root"
>
  <div
    className="container"
  >
    <div>
      <div
        className="headerContainer"
      >
        <div
          className="header"
        />
      </div>
      <main
        className="mainContainer"
      >
        <h2
          className="viewTitle"
        >
          View Title
        </h2>
        <div
          className="subtitleContainer"
        >
          View Subtitle
        </div>
      </main>
      <div
        className="footerContainer"
      >
        <div
          className="footer"
        />
      </div>
    </div>
  </div>
</div>
`;
